crypto/bn/bn_exp.c

```diff
@@ -1390,3 +1390,86 @@ int BN_mod_exp_simple(BIGNUM *r, const BIGNUM *a, const BIGNUM *p,
     bn_check_top(r);
     return ret;
 }
+
+/*
+ * This is a variant of modular exponentiation optimization that does
+ * parallel 2-primes exponentiation using 256-bit (AVX512VL) AVX512_IFMA ISA
+ * in 52-bit binary redundant representation.
+ * If such instructions are not available it falls back to two
+ * BN_mod_exp_mont_consttime() calls.
+ */
+int BN_mod_exp_mont_consttime_x2(BIGNUM *rr1, const BIGNUM *a1, const BIGNUM *p1,
+                                 const BIGNUM *m1, BN_MONT_CTX *in_mont1,
+                                 BIGNUM *rr2, const BIGNUM *a2, const BIGNUM *p2,
+                                 const BIGNUM *m2, BN_MONT_CTX *in_mont2,
+                                 BN_CTX *ctx)
+{
+    int ret = 0;
+
+#ifdef RSAZ_ENABLED
+    BN_MONT_CTX *mont1 = NULL;
+    BN_MONT_CTX *mont2 = NULL;
+
+    if (rsaz_avx512ifma_eligible() &&
+        ((16 == a1->top) && (16 == p1->top) && (BN_num_bits(m1) == 1024) &&
+         (16 == a2->top) && (16 == p2->top) && (BN_num_bits(m2) == 1024))) {
+
+        if (NULL == bn_wexpand(rr1, 16))
+            goto err;
+        if (NULL == bn_wexpand(rr2, 16))
+            goto err;
+
+        /*  Ensure that montgomery contexts are initialized */
+        if (NULL != in_mont1)
+            mont1 = in_mont1;
+        else {
+            if ((mont1 = BN_MONT_CTX_new()) == NULL)
+                goto err;
+            if (!BN_MONT_CTX_set(mont1, m1, ctx))
+                goto err;
+        }
+        if (NULL != in_mont2)
+            mont2 = in_mont2;
+        else {
+            if ((mont2 = BN_MONT_CTX_new()) == NULL)
+                goto err;
+            if (!BN_MONT_CTX_set(mont2, m2, ctx))
+                goto err;
+        }
+
+        /* Currently only 2048-bit moduli is supported. */
+        ret = RSAZ_mod_exp_avx512_x2(rr1->d, a1->d, p1->d, m1->d, mont1->RR.d,
+                                     mont1->n0[0],
+                                     rr2->d, a2->d, p2->d, m2->d, mont2->RR.d,
+                                     mont2->n0[0],
+                                     1024 /* factor bit size */);
+
+        rr1->top = 16;
+        rr1->neg = 0;
+        bn_correct_top(rr1);
+        bn_check_top(rr1);
+
+        rr2->top = 16;
+        rr2->neg = 0;
+        bn_correct_top(rr2);
+        bn_check_top(rr2);
+
+        goto err;
+    }
+#endif
+
+    /* m1 = m1^dmq1 mod q */
```

Comment:
I'm struggling to relate the names in this comment to the variable names below? Similarly for the other similar line

Reply:
Oops, copy-paste... Fixed.